src/o365/graph/commands/teams/teams-clone.spec.ts

```diff
@@ -0,0 +1,356 @@
+    const actual = (command.validate() as CommandValidate)({
+      options: {
+        teamId: '6f6fd3f7-9ba5-4488-bbe6-a789004d0d55',
+        displayName: "Library Assist",
+        partsToClone: "apps,tabs,settings,channels,members"
+      }
+    });
+    assert.notEqual(actual, true);
+    done();
+  });
+
+  it('fails validation if the partsToClone is not provided.', (done) => {
+    const actual = (command.validate() as CommandValidate)({
+      options: {
+        teamId: '6f6fd3f7-9ba5-4488-bbe6-a789004d0d55',
+        displayName: "Library Assist",
+        mailNickname: "libassist"
+      }
+    });
+    assert.notEqual(actual, true);
+    done();
+  });
+
+  it('fails validation on invalid visibility', () => {
+    const actual = (command.validate() as CommandValidate)({ options: { visibility: 'abc' } });
+    assert.notEqual(actual, true);
+  });
+
+  it('passes validation on valid \'private\' visibility', () => {
+    const actual = (command.validate() as CommandValidate)({
+      options: {
+        teamId: '15d7a78e-fd77-4599-97a5-dbb6372846c5',
+        displayName: "Library Assist",
+        mailNickname: "libassist",
+        partsToClone: "apps,tabs,settings,channels,members",
+        visibility: 'private'
+      }
+    });
+    assert.equal(actual, true);
+  });
+
+  it('passes validation on valid \'public\' visibility', () => {
+    const actual = (command.validate() as CommandValidate)({
+      options: {
+        teamId: '15d7a78e-fd77-4599-97a5-dbb6372846c5',
+        displayName: "Library Assist",
+        mailNickname: "libassist",
+        partsToClone: "apps,tabs,settings,channels,members",
+        visibility: 'public'
+      }
+    });
+    assert.equal(actual, true);
+  });
+
+  it('passes validation when the input is correct with mandatory parameters', () => {
+    const actual = (command.validate() as CommandValidate)({
+      options: {
+        teamId: '15d7a78e-fd77-4599-97a5-dbb6372846c5',
+        displayName: "Library Assist",
+        mailNickname: "libassist",
+        partsToClone: "apps,tabs,settings,channels,members"
+      }
+    });
+    assert.equal(actual, true);
+  });
+
+  it('passes validation when the input is correct with mandatory and optional parameters', () => {
+    const actual = (command.validate() as CommandValidate)({
+      options: {
+        teamId: '15d7a78e-fd77-4599-97a5-dbb6372846c5',
+        displayName: "Library Assist",
+        mailNickname: "libassist",
+        partsToClone: "apps,tabs,settings,channels,members",
+        description: "Self help community for library",
+        visibility: "public",
+        classification: "public"
+      }
+    });
+    assert.equal(actual, true);
+  });
+
+  it('creates a copy of a Microsoft Teams team with mandatory parameters', (done) => {
+    sinon.stub(request, 'post').callsFake((opts) => {
+      if (opts.url === `https://graph.microsoft.com/v1.0/teams/f5dba91d-6494-4d5e-89a7-ad832f6946d6/clone`) {
```

Comment:
@sprider , your stub has to fake response based on request with url where the team id is `15d7a78e-fd77-4599-97a5-dbb6372846c5` instread of `f5dba91d-6494-4d5e-89a7-ad832f6946d6` since you are calling the command with option of `teamId=15d7a78e-fd77-4599-97a5-dbb6372846c5`. So you you change that line 
```
if (opts.url === `https://graph.microsoft.com/v1.0/teams/f5dba91d-6494-4d5e-89a7-ad832f6946d6/clone`) {
...
``` 
to 
```
if (opts.url === `https://graph.microsoft.com/v1.0/teams/15d7a78e-fd77-4599-97a5-dbb6372846c5/clone`) {
...
```
then you will be able to fake the response for the request that your command will call.
When you crate a stub, by specifying the `opts.url === 'https://someurl'` the stub listens for post requests that are made with that url and then performs what is under the if block.

An easy way to debug your stub and how it behaves is to use simply `console.log`.
You can modify your package.json to run only your command tests for quick execution like:
![image](https://user-images.githubusercontent.com/1049476/56692207-94174480-66d9-11e9-8f2c-3fc884e8f22a.png)

Then place a `console.log` when you think you can get more information.   

![image](https://user-images.githubusercontent.com/1049476/56692278-c0cb5c00-66d9-11e9-8722-83c87c7ad42f.png)

Then you can run your tests and see the output from `console.log`.

![image](https://user-images.githubusercontent.com/1049476/56692141-6d590e00-66d9-11e9-9205-96a291ee1ee9.png)

As you can see the request is being called with teamId  `15d7a78e-fd77-4599-97a5-dbb6372846c5` therefore it will not enter in the if block, because the request url is different. 

I hope that makes sense. Let me know if you have more questions.

Reply:
@VelinGeorgiev,
Perfect, Thank you for your valuable time and feedbacks regarding debugging. I will keep in mind.